src/goto-programs/remove_java_new.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright 2017 Diffblue Limited. All Rights Reserved.
+
+/// \file
+/// Function-level/module-level pass to remove java_new and replace with
+/// malloc & zero-initialize
+
+#include "remove_java_new.h"
+
+#include <util/pointer_offset_size.h>
+#include <linking/zero_initializer.h>
+
+#include "class_identifier.h"
+
+static bool remove_java_new(
+  goto_programt &goto_program,
+  const namespacet &ns,
+  message_handlert &message_handler)
+{
+  messaget msg(message_handler);
+
+  bool changed=false;
+  for(goto_programt::targett target=goto_program.instructions.begin();
+    target!=goto_program.instructions.end();
+    ++target)
+  {
+    code_assignt *assign=expr_try_dynamic_cast<code_assignt>(target->code);
+    if(assign==nullptr)
+      continue;
+    side_effect_exprt *rhs=
+      expr_try_dynamic_cast<side_effect_exprt>(assign->rhs());
+    if(rhs==nullptr)
+      continue;
+    if(rhs->get_statement()!=ID_java_new)
+      continue;
+    INVARIANT(rhs->operands().empty(), "java_new does not have operands");
+    INVARIANT(rhs->type().id()==ID_pointer, "java_new returns pointer");
+
+    const exprt &lhs=assign->lhs();
+    INVARIANT(
+      !lhs.is_nil(), "remove_java_new without lhs is yet to be implemented");
+
+    typet object_type=rhs->type().subtype();
+
+    // build size expression
+    exprt object_size=size_of_expr(object_type, ns);
+    CHECK_RETURN(object_size.is_not_nil());
+
+    changed=true;
+
+    // We produce a malloc side-effect
+    side_effect_exprt malloc_expr(ID_allocate);
+    malloc_expr.copy_to_operands(object_size);
+    // could use true and get rid of the code below
+    malloc_expr.copy_to_operands(false_exprt());
```

Comment:
Somewhat unrelated: it may be time to do what the comment above says, at which point I'm really wondering whether there is a place for `java_new` at all?

This would likely also yield performance improvements as constant propagation on structs would start happening.